### PR TITLE
feat(webapp): let the deployment S2 client endpoints be overridden

### PR DESCRIPTION
**Pull request by d-cs**
## Summary

Deployment event logs always talked to hosted S2, so running the local `s2` container from `docker-compose` could not serve them: the deploy start path fails with a server error. A single optional env var now points that client somewhere else.

```
S2_DEPLOYMENT_ENDPOINT=http://localhost:4566
```

Unset means the SDK's hosted defaults, so nothing changes for existing deployments. The full local setup also needs `S2_ENABLED=1`, `S2_ACCESS_TOKEN`, and `S2_DEPLOYMENT_LOGS_BASIN_NAME=trigger-local` to match the basin `docker/config/s2-spec.json` creates.

## Design

The S2 SDK honours only the endpoints passed to its constructor. `S2Environment.parse()`, which reads endpoint variables from the environment, is an opt-in helper we never called, so the client fell back to `a.s2.dev` and `{basin}.b.s2.dev` regardless of configuration. Realtime streams already had an equivalent knob.

One value covers both the account and basin hosts on purpose. The SDK takes them separately, and setting only one leaves the other on the hosted default: the access token would go to hosted S2 over the public internet while the operator believed the client was entirely local, with nothing logged. A single value makes that state unrepresentable.

With the variable unset, the options object carries no `endpoints` key at all, so the call into the SDK is exactly the one production already makes. That distinction is not observable on the built client, because an `endpoints` key whose members are undefined resolves to the same hosted URLs, so the options builder is exported and asserted directly. Putting the key back on the unset path fails the test.

## Scope

This covers writing deployment event logs. Reading them back is still hosted-only: the dashboard hook and the CLI build-log tail construct their own clients, so log viewing against a local S2 does not work yet. Worth doing as a follow-up, since the server already mints a local token those readers cannot use.

Verified against the local container: a deploy completes and the event stream is created in the local basin.

### CODE REVIEW

File: apps/webapp/app/v3/s2Client.server.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  if (env.S2_ENABLED !== "1") {
+    return undefined;
+  }
+
+  const account = env.S2_ACCOUNT_ENDPOINT;
+  const basin = env.S2_BASIN_ENDPOINT;
+
+  return new S2({
+    accessToken: env.S2_ACCESS_TOKEN,
+    ...(account !== undefined || basin !== undefined ? { endpoints: { account, basin } } : {}),
```
[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/app/v3/s2Client.server.ts]: Good catch, acted on. The token does go to whatever endpoint is configured as a bearer header, so cleartext off the machine would leak it.

`S2_DEPLOYMENT_ENDPOINT` is now trimmed and refined: https to any host, http only to a loopback host (`localhost`, `127.0.0.1`, `[::1]`). A non-loopback http endpoint fails at boot naming the variable, so there is no configuration that quietly sends the token in the clear.

Worth recording why zod `.url()` was not enough on its own: it accepts `htp:/localhost:4566` (scheme `htp:`, empty host) and any plain `http://` host. Both now fail. `isValidS2Endpoint` is in `apps/webapp/app/utils/s2Endpoint.ts` with tests, and dropping the loopback restriction turns one of them red.

I did not take the separate non-privileged-token path. It is a larger change than this needs, and the s2-lite container accepts any token, so a local setup already need not use a real credential.

[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/app/v3/s2Client.server.ts]: Partly right, and I checked rather than assumed. The two variables this targets are gone (one `S2_DEPLOYMENT_ENDPOINT` now covers both hosts, so a half-set config cannot exist), but the whitespace half of the point stood: `z.string().url()` accepts `" http://localhost:4566"` with the leading space and passes it straight through.

The schema now `.trim()`s before validating, so that input normalises rather than reaching the SDK with a space in the host.

One deliberate difference from the suggestion: I did not fold whitespace-only into `undefined` beyond the empty-string case. `"   "` trims to `""` which becomes `undefined`, but a malformed non-empty value fails at boot naming the variable rather than being silently ignored. For a deployment-critical endpoint a loud failure beats a silent fallback to hosted S2.

File: apps/webapp/app/env.server.ts [hunk 2]
```diff
@@ -79,6 +79,15 @@ const S2EnvSchema = z.preprocess(
...
       S2_DEPLOYMENT_STREAMS_LOCAL: z.string().default("0"),
+      // Points deployment event logs at an S2 service other than the hosted one, e.g. the
+      // local s2-lite in docker compose. One value covers both the account and basin
+      // endpoints: splitting them lets a half-set config send the access token to the
+      // hosted service while the operator believes they are entirely local.
+      S2_DEPLOYMENT_ENDPOINT: z
+        .string()
+        .url()
+        .optional()
+        .or(z.literal("").transform(() => undefined)),
```
[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/app/env.server.ts]: Declining this one. Crumbs are temporary debug instrumentation that is stripped before merge, so adding them to a branch that is about to land would put markers into `main` that are meant never to get there. This was settled repo-wide previously, so it applies to every PR rather than just this one.

File: apps/webapp/test/s2Client.server.test.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import { describe, expect, it } from "vitest";
+import { buildDeploymentS2Client, deploymentS2ClientOptions } from "~/v3/s2Client.server";
```
[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/test/s2Client.server.test.ts]: Correct, and it cites a real guideline I had broken. Fixed.

The pure builders moved to `apps/webapp/app/v3/s2ClientConfig.ts`, which imports only the S2 SDK. `s2Client.server.ts` is now just the env-reading factory. The test imports the config module, so `env.server.ts` is no longer in its import graph, directly or transitively. That matches the split the guidelines point at, where a testable unit takes configuration and a separate `.server` module supplies it from the environment.

The test file is renamed to `s2ClientConfig.test.ts` to match what it now covers. `isValidS2Endpoint` and its test were already environment-free.

File: apps/webapp/app/v3/s2Client.server.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import { buildDeploymentS2Client } from "~/v3/s2ClientConfig";
+
+export function createDeploymentS2Client(): S2 | undefined {
+  if (env.S2_ENABLED !== "1") {
+    return undefined;
+  }
+
+  return buildDeploymentS2Client({
+    accessToken: env.S2_ACCESS_TOKEN,
+    endpoint: env.S2_DEPLOYMENT_ENDPOINT,
```
[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/app/v3/s2Client.server.ts:12]: Real bug, and I would not have found it. Fixed in `6c16d8b1b`.

Both caches were keyed by project alone, which was sufficient while the endpoint could never change. This PR makes it changeable, so Redis outliving a restart means a token issued by the previous S2 service is served for the rest of its hour, and log readers get credentials the new service will not accept.

Both keys now carry an endpoint scope: `s2CacheScope(env.S2_DEPLOYMENT_ENDPOINT)`, applied in `deployment.server.ts` and `DeploymentPresenter.server.ts`.

With no endpoint set the scope is the empty string, so hosted keys are byte-identical to what is in Redis today and nothing is invalidated where the endpoint cannot change. That is the same principle as the client construction itself: the unset path has to be exactly what production already does. A configured endpoint gets its own namespace, so switching endpoints is a cache miss rather than a stale token.

File: apps/webapp/app/utils/s2Endpoint.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    url = new URL(value);
+  } catch {
+    return false;
+  }
+
+  if (url.protocol === "https:") {
+    return true;
+  }
+
+  return url.protocol === "http:" && LOOPBACK_HOSTS.has(url.hostname);
```
[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/app/utils/s2Endpoint.ts]: Correct, and this would have broken the documented self-hosted path. Fixed in `6c16d8b1b`.

`hosting/docker/webapp/docker-compose.yml:54` already reaches S2 over http at a service name for realtime (`REALTIME_STREAMS_S2_ENDPOINT: http://s2/v1`), so requiring TLS here would have made the deployment endpoint the odd one out on the same private network.

The rule is now https anywhere, and http only to a host that is not reachable from the public internet: loopback, a single-label hostname (a container or service name, since it has no dot), or a private IPv4 range. `http://s2/v1` and `http://10.0.0.5:4566` are accepted; `http://s2.example.com`, `http://a.s2.dev` and `http://8.8.8.8` are not, so the original credential concern still holds.

`172.32.0.1` is in the tests as a boundary case, since it is outside the private `172.16` to `172.31` block and must be rejected.

File: apps/webapp/test/s2ClientConfig.test.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import { describe, expect, it } from "vitest";
+import { buildDeploymentS2Client, deploymentS2ClientOptions } from "~/v3/s2ClientConfig";
```
[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/test/s2ClientConfig.test.ts]: Accepted, moved to `apps/webapp/app/v3/s2ClientConfig.test.ts`.

The reason it was in `test/` is that vitest would not have collected it: the webapp config lists colocated paths explicitly rather than globbing all of `app/`. So the move needed the path registering too, alongside the existing `app/v3/runStore.server.test.ts` entry, which is the same situation.

`test/s2Endpoint.test.ts` and `test/s2CacheScope.test.ts` stay where they are, matching their siblings for `app/utils` and other non-colocated helpers.

File: apps/webapp/app/v3/s2ClientConfig.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  accessToken,
+  endpoint,
+}: DeploymentS2Config): DeploymentS2ClientOptions {
+  if (endpoint === undefined) {
+    return { accessToken };
+  }
+
+  // One value drives both hosts. Overriding just one would send the access token to the hosted
+  // service while the other half went elsewhere.
+  return { accessToken, endpoints: { account: endpoint, basin: endpoint } };
```
[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/app/v3/s2ClientConfig.ts:25]: Declining, as with the other crumbs threads. They are temporary debug instrumentation stripped before merge, so adding them to a branch about to land would put markers into `main` that are meant never to reach it. Settled repo-wide.

Noting the specific caution was right though: the access token must never appear in any diagnostic. It is not logged anywhere in these paths, and the validation error message names the variable and the rule rather than the value.

File: apps/webapp/app/env.server.ts [hunk 8]
```diff
@@ -79,6 +80,19 @@ const S2EnvSchema = z.preprocess(
...
+      // hosted service while the operator believes they are entirely local.
+      S2_DEPLOYMENT_ENDPOINT: z
+        .string()
+        .trim()
+        .transform((value) => (value === "" ? undefined : value))
+        .refine(
+          (value) => value === undefined || isValidS2Endpoint(value),
+          "must be an http(s) URL; http is only allowed for a loopback host, because the S2 access token is sent to it as a bearer token"
+        )
+        .optional(),
```
[the remark this answers is not in the document]
Comment (d-cs) [reply apps/webapp/app/env.server.ts:95]: Deliberately omitted, on a maintainer call rather than an oversight.

The judgement was that this is local and self-hosted configuration plumbing rather than something a cloud user would notice or act on, which is the bar `.server-changes` entries are meant to clear. Happy to add one if the view changes; it would be a single line naming the new variable.
